cpp/src/arrow/array/data.cc

```diff
@@ -286,8 +286,10 @@ namespace {
 template <typename offset_type>
 BufferSpan OffsetsForScalar(uint8_t* scratch_space, offset_type value_size) {
   auto* offsets = reinterpret_cast<offset_type*>(scratch_space);
-  offsets[0] = 0;
-  offsets[1] = static_cast<offset_type>(value_size);
+  reinterpret_cast<std::atomic<offset_type>*>(&offsets[0])
+      ->store(0, std::memory_order_relaxed);
+  reinterpret_cast<std::atomic<offset_type>*>(&offsets[1])
+      ->store(static_cast<offset_type>(value_size), std::memory_order_relaxed);
```

Comment:
I think we can reduce the number of reinterpret casts here
```suggestion
  auto* offsets = reinterpret_cast<std::atomic<offset_type>*>(scratch_space);
  offsets[0].store(0, std::memory_order_relaxed);
  offsets[1].store(value_size, std::memory_order_relaxed);
```

Reply:
Well, the reason for the extra reinterpret casts is that: I'm not sure if it is guaranteed in C++ that `sizeof(std::atomic<offset_type>)` is the same as `sizeof(offset_type)`. If they are not the same, the pointer will be wrongly offset-ed.

Reply:
You are right. The casts could be reduced. After a second thought, I think even with extra casts, the result would be wrong if the size of atomic is not the same as the payload. I tend to assume we are not going to support platforms that need a mutex for atomic 64. So I'll just reduce the casts. Thank you.